verification-by-reference.md

```diff
@@ -0,0 +1,331 @@
+  "validTo": "6/3/2020",
+  "x509-fingerprint": "43:51:43:a1:b5:fc:8b:b7:0a:3a:a9:b1:0f:66:73:a8"
+}
+```
+
+### web:a2b2 scanning verification w/ index
+
+Another index of type `notary.verification`, which adds the scanning verification.
+
+```json
+{
+  "schemaVersion": 2.1,
+  "config": {
+    "mediaType": "application/vnd.cncf.notary.verification.config.v1+json",
+    "size": 7023,
+    "digest": "sha256:a6a2b2c507a0944348e0303114d8d93aaaa081732b86451d9"
+  },
+  "manifests": [
+    {
+      "mediaType": "application/vnd.oci.image.manifest.v1+json",
+      "size": 7023,
+      "digest": "sha256:b5b2b2c507a0944348e0303114d8d93aaaa081732b86451",
+      "platform": {
+        "architecture": "ppc64le",
+        "os": "linux"
+      }
+    }
+  ]
+}
+```
+
+### web:a2b2 scanning verification-config
+
+```json
+{
+  "issuedTo": "*.scan.contoso.com",
+  "issuedBy": "DigCerti SHA2 Extended Validation Server CA",
+  "validFrom": "5/7/2018",
+  "validTo": "6/3/2020",
+  "x509-fingerprint": "86:75:30:90:b5:fc:8b:b7:0a:3a:a9:13:12:55:62:b8"
+}
+```
+
+The nv2 client could retrieve all signatures with the following mocked command
+
+```shell
+nv2 verification pull registry.contoso.com/marketing/web:a2b2
+nv2 verification pull registry.contoso.com/marketing/web@256:efg123
+```
+
+or
+
+```shell
+nv2 verify registry.contoso.com/marketing/web:a2b2
+nv2 verify registry.contoso.com/marketing/web@256:efg123
+```
+
+The `nv2 verify` API compares the artifact with local configuration to attest to it matching the configured rules. These could include the artifact has required verifications for Scanning and Testing verification.
+
+## Pros & Cons
+
+The pros of this approach are:
+
+- Multiple signatures may be added without changing the original artifact digest as no content changes are made to the digest, or even annotations
+- All deployment references remain: `web:a2b2` and it’s associated digest do not change.
+
+The cons of this approach:
+
+- A new API must be added. We’ve said this was within scope and ok as registries must add Notary v1, Content Trust to support existing scenarios. Since most registries need to do some work, adding a new API is not a stretch
+- Reverse lookup, and its impact on garbage collection. Registries tend to traverse down. If multiple manifests reference common base layers, the registry knows to reference count shared layers and only deletes the layers once all manifests are removed. Orphaned manifests may be deleted if they’re no longer referenced by tags.  
+Most importantly, a manifest is likely blocked from deletion if it’s referenced by an index. In this reverse lookup model, the user likely wants all verifications to be deleted when the artifact is deleted. This isn’t necessarily an immediate blocker, but it does go against the garbage collection constraint outlined above.
+This is work registry operators must implement so users don’t have to change their artifact references? Is this the right trade-off?
```

Comment:
I really struggle with this.

There are tons of registry implementations out there, many of which will never be updated. Any API that we add will cause some friction, and I'd be somewhat worried about incompatible implementations of that API.

If we can do everything client side without updating registries, we can more easily get widespread adoption without being blocked by registries. It would be nice if there was some fallback pattern for registries that don't implement a reverse lookup.

Reply:
Not having an impact on a registry to support Artifacts was the goal, as we knew it didn't hit the "high bar" for requiring a major change. We actually just require a registry operator to relax the mediaType constraints, which is consistent with the spec.
For Signing, we think this is much needed feature. Registries that implemented Notary v1/Docker Content Trust had a LOT of work to do. For notary v2, we're hoping this is limited to one or two APIs.

The point being, the desired results are worth the added investment. If a registry operator doesn't want to support signing, that's their choice. I suspect all registries will want/need to support signing, so we just need to identify the benefits for security and usability.